### PR TITLE
Block start mapping and start validating if email is not set

### PR DESCRIPTION
**Pull request by uwaiszaki**
This will block start mapping and start validating button with a warning message if email is not set

### CODE REVIEW

**Comment by uwaiszaki** [comment]
Not tested completely because of #1430

**Comment by ethan-nelson** [comment]
Thanks @uwaiszaki...there is a backup right now due to deployment/docker. We should hopefully have it updated soon.

**Comment by xamanu** [comment]
Thanks for the contribution! The current problem on the `develop` branch should not effect this here. I also gave some feedback in #1430 to get you back on track, @uwaiszaki.

File: client/app/project/project.controller.js [hunk 1]
```diff
@@ -1266,35 +1275,39 @@
...
-                vm.validatingStep = 'locked';
-                vm.selectedTaskData = tasks[0];
-                vm.isSelectedValidatable = true;
-                vm.lockedTaskData = tasks[0];
-                vm.lockTime[taskId] = getLastLockedAction(vm.lockedTaskData).actionDate;
-                formatHistoryComments(vm.selectedTaskData.taskHistory);
-            }, function (error) {
-                onLockError(projectId, error);
-            });
+            if(vm.user.emailAddress != 'uwaiszaki104@gmail.com'){
```
Comment (willemarcel) [inline client/app/project/project.controller.js]: Hi @uwaiszaki,
Could you improve the code to make it available to any user. So you should verify here if the username email is not null or `''`.

Comment (uwaiszaki) [reply client/app/project/project.controller.js]: @willemarcel I have added the changes and also rebased this branch in the develop branch.

**Comment by uwaiszaki** [comment]
@xamanu @willemarcel On which step should I add backend email restriction?

**Comment by xamanu** [comment]
@fitoria can you please give some guidance? Thanks.

**Comment by fitoria** [comment]
This works fine but I'm wondering if it's better to check against the is_email_verified field instead of email_address, specially to avoid spam concerns, in a related matter can we also validate the project questions and comments form? Or put a captcha if the email is not set there.

And on the error message where it says: "Add Your Email To Continue Mapping" it should have a link to the edit profile form so it should be like "<a href="ediprofilelink" >Add Your Email</a> To Continue Mapping"

**Comment by uwaiszaki** [comment]
@xamanu @fitoria I have added the requested changes but the "Add Your Email" Link is giving an odd look with the error message.
tmblockfeatureissue

File: client/app/project/project.html [hunk 2]
```diff
@@ -562,13 +566,17 @@ <h3>{{ 'Validation' | translate }}</h3>
                                             class="alert alert--warning"
                                             role="alert">
                                         <p>In order to start contributing, please <a ng-show="!accountNavCtrl.account.username" ng-click="accountNavCtrl.login()">login first</a>.</p>
-                                    </div>
+                                    </div>0
```
Comment (xamanu) [inline client/app/project/project.html]: Is this `0` on purpose?

File: client/app/project/project.html [hunk 3]
```diff
@@ -371,6 +371,10 @@ <h3 style="display: inline-block;">{{ 'Mapping' | translate }}</h3>
                                         <!-- task action buttons -->
                                         <div class="form__group form__group--section">
                                             <div>
+                                                <div ng-show="projectCtrl.showWarning">
+                                                    <p class="error"><a class='emailWarning' href="/user/{{projectCtrl.user.username}}">{{ 'Add Your Email' | translate }}</a>
+                                                        {{ ' To Continue Mapping' | translate }}</p>
```
Comment (xamanu) [inline client/app/project/project.html]: I suggest to use `{{ 'to continue validation' }}` and put the white space before the translatable string. I also think we can go with normal not capitalized words. "Add your email address" "to continue validation".

File: client/app/project/project.html [hunk 4]
```diff
@@ -562,13 +566,17 @@ <h3>{{ 'Validation' | translate }}</h3>
...
+                                    </div>0
                                     <div ng-show="projectCtrl.user && !projectCtrl.userCanValidate"
                                             class="alert alert--warning"
                                             role="alert">
                                         <p>{{ 'You cannot select tasks for validation on this project because you do not have the Validator role.' | translate }}</p>
                                     </div>
-                                    <div ng-show="projectCtrl.userCanValidate">
+                                    <div ng-show="projectCtrl.showWarning">
+                                            <p class="error"><a class='emailWarning' href="/user/{{projectCtrl.user.username}}">{{'Add Your Email'| translate}}</a>
+                                                {{ ' To Continue Validation' | translate }}</p>
```
Comment (xamanu) [inline client/app/project/project.html]: Same here.

File: client/app/project/project.html [hunk 5]
```diff
@@ -749,6 +757,10 @@ <h3 style="display: inline-block;">{{ 'Validation' | translate }}</h3>
                                     <!-- task action buttons -->
                                     <div class="form__group form__group--section">
                                         <div>
+                                            <div ng-show="projectCtrl.showWarning">
+                                                <p class="error"><a class='emailWarning' href="/user/{{projectCtrl.user.username}}">{{'Add Your Email'| translate}}</a>
+                                                    {{ ' To Continue Validation' | translate }}</p>
```
Comment (xamanu) [inline client/app/project/project.html]: Same here.

File: client/app/project/project.html [hunk 6]
```diff
@@ -777,6 +789,7 @@ <h3 style="display: inline-block;">{{ 'Validation' | translate }}</h3>
                                                     {{ 'Cancel' | translate }}
                                                 </button> -->
                                             </div>
+
```
Comment (xamanu) [inline client/app/project/project.html]: Is this on purpose?

File: client/app/project/project.html [hunk 7]
```diff
@@ -949,8 +962,14 @@ <h6>{{ 'Done editing? Leave a comment and select one of the options below that m
...
                         <div ng-show="projectCtrl.currentTab === 'chat'">
                             <h3>{{ 'Project Questions and Comments' | translate }}</h3>
-                            <project-chat project-id="projectCtrl.id"
+                            <div ng-show="projectCtrl.user.isEmailVerified">
+                                <project-chat project-id="projectCtrl.id"
                                             project-author="projectCtrl.projectData.author"></project-chat>
+                            </div>
+                            <div class='error' ng-hide="projectCtrl.user.isEmailVerified">
+                                <a class='emailWarning' href="/user/{{ projectCtrl.user.username }}">{{ 'Add Your Email' | translate }}</a> 
+                                {{ 'To Continue' | translate }}
```
Comment (xamanu) [inline client/app/project/project.html]: Same here with capitalized words.

File: client/assets/styles/sass/_project.scss [hunk 8]
```diff
@@ -23,6 +23,10 @@
     font-size: 0.90rem;
   }
 
+  .emailWarning {
```
Comment (xamanu) [inline client/assets/styles/sass/_project.scss]: Can't we use the same formatting as we use for other error messages with `.error`?

Comment (uwaiszaki) [reply client/assets/styles/sass/_project.scss]: @xamanu I have used this to highlight the link. What do u think?  Thanks

**Comment by uwaiszaki** [comment]
@xamanu Can u give ur reviews on the use of color on the link for the edit profile page?

**Comment by xamanu** [comment]
i would just make the message clickable (link) without adding any other color to symbolize a link to it.

**Pull request closed without merging**